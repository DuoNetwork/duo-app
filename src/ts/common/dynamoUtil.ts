import { DynamoUtil } from '@finbook/duo-market-data';

let config = {};
try {
	config = require(`../keys/aws.ui.${__KOVAN__ ? 'dev' : 'live'}.json`);
} catch (error) {
	console.log(error);
}

const dynamoUtil = new DynamoUtil(config, !__KOVAN__, '');

export default dynamoUtil;
