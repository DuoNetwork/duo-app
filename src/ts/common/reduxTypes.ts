import { IAssets, IPriceData, ITimeSeriesData } from './types';
export interface IState {
	ui: IUIState;
}

export interface IUIState {
	eth: ITimeSeriesData[];
	classA: ITimeSeriesData[];
	classB: ITimeSeriesData[];
	reset: ITimeSeriesData[];
	history: string[];
	message: {
		type: string;
		content: string;
		visible: boolean;
	};
	setting: {
		couponRate: number;
		upwardResetLimit: number;
		downwardResetLimit: number;
		periodicResetLimit: number;
	};
	form: {
		type: string;
		visible: boolean;
	};
	mv: ITimeSeriesData[];
	assets: IAssets;
	resetPrice: number;
	beta: number;
	day: number;
	upward: number;
	downward: number;
	periodic: number;
	price: IPriceData;
}

export type Action =
	| IBaseAction
	| IBooleanAction
	| IStringAction
	| IObjectAction
	| INumberAction
	| ITradeAction;

export type Dispatch = (action: Action | ThunkAction | PromiseAction) => any;

export interface IBaseAction {
	type: string;
}

export interface IBooleanAction {
	type: string;
	value: boolean;
}

export interface IStringAction {
	type: string;
	value: string;
}

export interface IObjectAction {
	type: string;
	value: object;
}

export interface INumberAction {
	type: string;
	value: number;
}

export interface ITradeAction {
	type: string;
	history: string;
	assets: IAssets;
}

export type ThunkAction = (dispatch: Dispatch) => any;
export type PromiseAction = Promise<Action>;
