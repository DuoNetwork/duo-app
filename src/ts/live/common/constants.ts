export * from '../../../../../duo-admin/src/constants';

export const AC_CTD_STATES = 'custodianStates';
export const AC_CTD_PRICES = 'custodianPrices';
export const AC_ACCOUNT = 'account';
export const AC_BALANCES = 'balances';
export const AC_ADDRESSES = 'addresses';
export const AC_DNM_STATUS = 'dynamoStatus';
export const AC_DMN_HOURLY = 'dynamoHourly';
export const AC_DMN_MINUTELY = 'dynamoMinutely';
export const AC_DMN_PRICES = 'dynamoPrices';
export const AC_REFRESH = 'refresh';
