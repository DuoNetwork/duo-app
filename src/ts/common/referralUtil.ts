import { AttributeMap } from 'aws-sdk/clients/dynamodb';
import moment from 'moment';
import dynamoUtil from './dynamoUtil';
import * as StakingCST from './stakingCST';
import { IAddressInfo, IReferral, IStakingChild, IStakingNode } from './types';
class ReferralUtil {
	public getUTCNowTimestamp() {
		return moment().valueOf();
	}

	public async checkExist(account: string) {
		const data = await dynamoUtil.scanData({ TableName: StakingCST.REFERRALTABLE });
		const addressList: string[] = [];
		if (data.Items)
			data.Items.map(item => {
				addressList.push((item as any).address.S);
			});
		if (addressList.includes(account)) {
			const index = addressList.indexOf(account);
			return (data.Items as any)[index].referralCode.S;
		}
		return '';
	}

	public async insertReferralEntry(item: IReferral) {
		const data: AttributeMap = {
			address: { S: item.address },
			referralCode: { S: item.referralCode },
			signHash: { S: item.signHash },
			updatedAt: { S: this.getUTCNowTimestamp() + '' }
		};
		const params = {
			TableName: StakingCST.REFERRALTABLE,
			Item: data
		};

		await dynamoUtil.insertData(params);
	}

	public async getAddressInfo(account: string) {
		const params = {
			TableName: StakingCST.STAKINGTABLE,
			KeyConditionExpression: 'address = :address',
			ExpressionAttributeValues: {
				[':address']: { S: account }
			}
		};
		const data = await dynamoUtil.queryData(params);
		if (data.Count) {
			const rawRA = (data.Items as any)[0].referralAward;
			const rawS0 = (data.Items as any)[0].staking0Award;
			const rawS60 = (data.Items as any)[0].staking60Award;
			const addressInfo: IAddressInfo = {};
			if (rawRA)
				if ((rawRA as any).S !== 'no referee') {
					const rawChildList = rawRA.S.split(';');
					const childList: IStakingChild[] = [];
					rawChildList.forEach((child: string) => {
						const rawData = child.split(',');
						const childData = {
							address: rawData[0],
							daily: Number(rawData[1]),
							accumulated: Number(rawData[2])
						};
						childList.push(childData);
					});
					addressInfo.children = childList;
				}
			if (rawS0) {
				const rawNodeList = rawS0.S.split(';');
				const nodeList: IStakingNode[] = [];
				rawNodeList.forEach((node: string) => {
					const rawData = node.split(',');
					const nodeData = {
						name: rawData[0],
						daily: Number(rawData[1]),
						accumulated: Number(rawData[2])
					};
					nodeList.push(nodeData);
				});
				addressInfo.staking0 = nodeList;
			}
			if (rawS60) {
				const rawNodeList = rawS60.S.split(';');
				const nodeList: IStakingNode[] = [];
				rawNodeList.forEach((node: string) => {
					const rawData = node.split(',');
					const nodeData = {
						name: rawData[0],
						daily: Number(rawData[1]),
						accumulated: Number(rawData[2])
					};
					nodeList.push(nodeData);
				});
				addressInfo.staking60 = nodeList;
			}
			return addressInfo;
		} else return {};
	}
}

const referralUtil = new ReferralUtil();
export default referralUtil;
