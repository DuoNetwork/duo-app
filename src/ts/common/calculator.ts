import moment from 'moment';
import * as CST from './constants';
import { IUIState } from './reduxTypes';
import { IAssets, IRawData, ITimeSeriesData } from './types';

class Calculator {
	public getAllTimeSeriesFromEth(
		rawData: IRawData[],
		alpha: number = 1,
		coupon: number = 0.0002,
		upperLimit: number = 2,
		lowerLimit: number = 0.25,
		periodLimit: number = 1.02
	) {
		const eth: ITimeSeriesData[] = [],
			classA: ITimeSeriesData[] = [],
			classB: ITimeSeriesData[] = [],
			resetPrice: ITimeSeriesData[] = [],
			beta: ITimeSeriesData[] = [],
			upward: ITimeSeriesData[] = [],
			downward: ITimeSeriesData[] = [],
			periodic: ITimeSeriesData[] = [];
		let count = -1;
		let lastResetDatetime = moment(rawData[0].Date, 'YYYY-MM-DD').valueOf();
		let lastResetPrice = rawData[0].Price;
		let currentBeta = 1;
		rawData.forEach(d => {
			count++;
			const datetime = moment(d.Date, 'YYYY-MM-DD').valueOf();
			const ethPx = d.Price;
			eth.push({
				datetime: datetime,
				value: ethPx
			});

			let [navA, navB] = this.calculateNav(
				ethPx,
				lastResetPrice,
				alpha,
				currentBeta,
				count,
				coupon
			);

			if (navB >= upperLimit || navB <= lowerLimit || navA >= periodLimit) {
				lastResetPrice = ethPx;
				lastResetDatetime = datetime;
				count = 0;
				if (navB >= upperLimit)
					upward.push({
						datetime: datetime,
						value: 1
					});
				else if (navB <= lowerLimit)
					downward.push({
						datetime: datetime,
						value: 1
					});
				else
					periodic.push({
						datetime: datetime,
						value: 1
					});

				// class B does not reset in periodic reset
				if (navA < periodLimit) {
					classB.push({
						datetime: datetime,
						value: navB
					});
					navB = 1;
					currentBeta = 1;
				} else
					currentBeta = this.updateBeta(currentBeta, ethPx, lastResetPrice, navA, alpha);

				classA.push({
					datetime: datetime,
					value: navA
				});

				navA = 1;
			}

			classA.push({
				datetime: datetime,
				value: navA
			});

			classB.push({
				datetime: datetime,
				value: navB
			});

			resetPrice.push({
				datetime: lastResetDatetime,
				value: lastResetPrice
			});

			beta.push({
				datetime: datetime,
				value: currentBeta
			});
		});
		return [eth, classA, classB, resetPrice, beta, upward, downward, periodic];
	}

	public calculateNav(
		price: number,
		resetPrice: number,
		alpha: number,
		beta: number,
		periodCount: number,
		coupon: number
	) {
		let navParent = price / resetPrice / beta * (1 + alpha);

		let navA = 1 + periodCount * coupon;
		let navAAdj = navA * alpha;
		if (navParent <= navAAdj) return [navParent / alpha, 0];
		else return [navA, navParent - navAAdj];
	}

	public updateBeta(
		prevBeta: number,
		price: number,
		resetPrice: number,
		navA: number,
		alpha: number
	) {
		return (
			(1 + alpha) * price / ((1 + alpha) * price - resetPrice * alpha * prevBeta * (navA - 1))
		);
	}

	public assetUpwardReset(assets: IAssets, eth: number, navA: number, navB: number): IAssets {
		return {
			ETH: assets.ETH + ((navA - 1) * assets.ClassA + (navB - 1) * assets.ClassB) / eth,
			ClassA: assets.ClassA,
			ClassB: assets.ClassB
		};
	}

	public assetDownwardReset(assets: IAssets, eth: number, navA: number, navB: number): IAssets {
		return {
			ETH: assets.ETH + (navA - navB) * assets.ClassA / eth,
			ClassA: assets.ClassA * navB,
			ClassB: assets.ClassB * navB
		};
	}

	public assetPeriodicReset(assets: IAssets, eth: number, navA: number): IAssets {
		return {
			ETH: assets.ETH + (navA - 1) * assets.ClassA / eth,
			ClassA: assets.ClassA,
			ClassB: assets.ClassB
		};
	}

	public calculateNextDayState(state: IUIState): object {
		const { eth, classA, classB, assets, day, upward, downward, periodic, mv } = state;
		const newDayCount = day + 1;
		if (newDayCount >= eth.length) return {};
		else {
			const newDatetime = eth[newDayCount].datetime;
			const newEthPx = eth[newDayCount].value;
			const upwardSoFar = upward.filter(d => d.datetime <= newDatetime);
			const downwardSoFar = downward.filter(d => d.datetime <= newDatetime);
			const periodicSoFar = periodic.filter(d => d.datetime <= newDatetime);
			const isUpward =
				upwardSoFar.length && upwardSoFar[upwardSoFar.length - 1].datetime === newDatetime;
			const isDownward =
				downwardSoFar.length &&
				downwardSoFar[downwardSoFar.length - 1].datetime === newDatetime;
			const isPeriodic =
				periodicSoFar.length &&
				periodicSoFar[periodicSoFar.length - 1].datetime === newDatetime;
			const upwardCount = upwardSoFar.length - (isUpward ? 1 : 0);
			const downwardCount = downwardSoFar.length - (isDownward ? 1 : 0);
			const periodicCount = periodicSoFar.length - (isPeriodic ? 1 : 0);
			const newNavA = classA[newDayCount + upwardCount + downwardCount + periodicCount].value;
			const newNavB = classB[newDayCount + upwardCount + downwardCount].value;

			let newAssets: IAssets;
			let msg = '';
			if (isUpward) {
				newAssets = calculator.assetUpwardReset(assets, newEthPx, newNavA, newNavB);
				msg = "<div style='color: rgba(255,255,255, .8)'>Reset (upward) triggered.</div>";
			} else if (isDownward) {
				newAssets = calculator.assetDownwardReset(assets, newEthPx, newNavA, newNavB);
				msg = "<div style='color: rgba(255,255,255, .8)'>Reset (downward) triggered.</div>";
			} else if (isPeriodic) {
				newAssets = calculator.assetPeriodicReset(assets, newEthPx, newNavA);
				msg = "<div style='color: rgba(255,255,255, .8)'>Reset (periodic) triggered.</div>";
			} else newAssets = assets;

			return Object.assign(
				{
					[CST.AC_MV]: [
						...mv,
						{
							datetime: newDatetime,
							value:
								assets.ETH * newEthPx +
								assets.ClassA * newNavA +
								assets.ClassB * newNavB
						}
					],
					[CST.AC_ASSETS]: newAssets,
					[CST.AC_DAY]: newDayCount,
					[CST.AC_PRICE]: {
						Date: newDatetime,
						ETH: newEthPx,
						ClassA:
							classA[
								newDayCount +
									upwardSoFar.length +
									downwardSoFar.length +
									periodicSoFar.length
							].value,
						ClassB:
							classB[newDayCount + upwardSoFar.length + downwardSoFar.length].value
					}
				},
				msg
					? {
							[CST.AC_MESSAGE]: {
								type: "<div style='color: rgba(0,186,255,0.7)'>INFORMATION</div>",
								content: msg,
								visible: true
							}
					}
					: {}
			);
		}
	}

	public calculateForwardState(state: IUIState): object {
		const { eth, classA, classB, assets, day, upward, downward, periodic, mv } = state;
		const currentDatetime = eth[day].datetime;
		const endDatetime = eth[eth.length - 1].datetime;
		const upwardSoFar = upward.filter(d => d.datetime <= currentDatetime);
		const downwardSoFar = downward.filter(d => d.datetime <= currentDatetime);
		const periodicSoFar = periodic.filter(d => d.datetime <= currentDatetime);
		const nextUpward = upward[upwardSoFar.length]
			? upward[upwardSoFar.length].datetime
			: endDatetime;
		const nextDownward = downward[downwardSoFar.length]
			? downward[downwardSoFar.length].datetime
			: endDatetime;
		const nextPeriodic = periodic[periodicSoFar.length]
			? periodic[periodicSoFar.length].datetime
			: endDatetime;
		const nextResetDate = Math.min(nextUpward, nextDownward, nextPeriodic);
		let curDate = day + 1;
		let newMV = [...mv];
		let newDatetime = eth[curDate].datetime;
		if (newDatetime === nextResetDate) return this.calculateNextDayState(state);

		let newEthPx, newNavA, newNavB;
		while (curDate < eth.length && eth[curDate].datetime < nextResetDate) {
			newDatetime = eth[curDate].datetime;
			newEthPx = eth[curDate].value;
			newNavA =
				classA[curDate + upwardSoFar.length + downwardSoFar.length + periodicSoFar.length]
					.value;
			newNavB = classB[curDate + upwardSoFar.length + downwardSoFar.length].value;
			newMV.push({
				datetime: newDatetime,
				value: assets.ETH * newEthPx + assets.ClassA * newNavA + assets.ClassB * newNavB
			});
			curDate++;
		}
		curDate--;
		return {
			[CST.AC_MV]: newMV,
			[CST.AC_DAY]: curDate,
			[CST.AC_PRICE]: {
				Date: newDatetime,
				ETH: newEthPx,
				ClassA:
					classA[
						curDate + upwardSoFar.length + downwardSoFar.length + periodicSoFar.length
					].value,
				ClassB: classB[curDate + upwardSoFar.length + downwardSoFar.length].value
			}
		};
	}
}

const calculator = new Calculator();
export default calculator;
