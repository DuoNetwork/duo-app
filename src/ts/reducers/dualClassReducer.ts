import { Constants as WrapperConstants } from '@finbook/duo-contract-wrapper';
import { AnyAction } from 'redux';
import * as CST from 'ts/common/constants';
import { IDualClassState } from 'ts/common/types';

export const initialState: IDualClassState = {
	states: {
		lastOperationTime: 0,
		operationCoolDown: 0,
		state: WrapperConstants.CTD_LOADING,
		minBalance: 0,
		totalSupplyA: 0,
		totalSupplyB: 0,
		collateral: 0,
		navA: 0,
		navB: 0,
		lastPrice: 0,
		lastPriceTime: 0,
		resetPrice: 0,
		resetPriceTime: 0,
		createCommRate: 0,
		redeemCommRate: 0,
		period: 0,
		maturity: 0,
		preResetWaitingBlocks: 0,
		priceFetchCoolDown: 0,
		nextResetAddrIndex: 0,
		totalUsers: 0,
		feeBalance: 0,
		resetState: '',
		alpha: 0,
		beta: 0,
		periodCoupon: 0,
		limitPeriodic: 0,
		limitUpper: 0,
		limitLower: 0,
		iterationGasThreshold: 0
	},
	addresses: {
		operator: WrapperConstants.DUMMY_ADDR,
		feeCollector: WrapperConstants.DUMMY_ADDR,
		roleManager: WrapperConstants.DUMMY_ADDR,
		oracle: WrapperConstants.DUMMY_ADDR,
		aToken: WrapperConstants.DUMMY_ADDR,
		bToken: WrapperConstants.DUMMY_ADDR
	},
	exchangePrices: [],
	acceptedPrices: [],
	conversions: [],
	balances: {
		a: 0,
		b: 0,
		duo: 0
	},
	type: WrapperConstants.BEETHOVEN,
	tenor: WrapperConstants.TENOR_PPT,
	subscription: 0
};

export function dualClassReducer(
	state: IDualClassState = initialState,
	action: AnyAction
): IDualClassState {
	switch (action.type) {
		case CST.AC_DCC_STATES:
			return Object.assign({}, state, {
				states: action.value
			});
		case CST.AC_DCC_ADDRESSES:
			return Object.assign({}, state, {
				addresses: action.value
			});
		case CST.AC_DCC_EX_PX:
			return Object.assign({}, state, {
				exchangePrices: action.value
			});
		case CST.AC_DCC_ACCEPTED_PX:
			return Object.assign({}, state, {
				acceptedPrices: action.value
			});
		case CST.AC_DCC_CONVERSIONS:
			return Object.assign({}, state, {
				conversions: action.value
			});
		case CST.AC_DCC_BALANCES:
			return Object.assign({}, state, {
				balances: action.value
			});
		case CST.AC_DCC_SUB:
			if (action.id)
				return Object.assign({}, state, {
					type: action.custodianType,
					tenor: action.tenor,
					subscription: action.id
				});
			else {
				window.clearInterval(state.subscription);
				return Object.assign({}, initialState, {
					type: action.custodianType,
					tenor: action.tenor
				});
			}
		default:
			return state;
	}
}
