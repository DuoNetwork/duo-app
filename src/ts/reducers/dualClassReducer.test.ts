// fix for @ledgerhq/hw-transport-u2f 4.28.0
import '@babel/polyfill';
import * as CST from 'ts/common/constants';
import { dualClassReducer, initialState } from './dualClassReducer';

describe('dualClass reducer', () => {
	let state = initialState;

	test('default', () => {
		state = dualClassReducer(state, { type: 'any' });
		expect(state).toMatchSnapshot();
	});

	test('dualClassStates', () => {
		state = dualClassReducer(state, {
			type: CST.AC_DCC_STATES,
			value: {
				test: 'test'
			}
		});
		expect(state).toMatchSnapshot();
	});

	test('dualClassAddresses', () => {
		state = dualClassReducer(state, {
			type: CST.AC_DCC_ADDRESSES,
			value: {
				test: 'test'
			}
		});
		expect(state).toMatchSnapshot();
	});

	test('dualClassExchangePrices', () => {
		state = dualClassReducer(state, {
			type: CST.AC_DCC_EX_PX,
			value: {
				test: 'test'
			}
		});
		expect(state).toMatchSnapshot();
	});

	test('dualClassAcceptedPrices', () => {
		state = dualClassReducer(state, {
			type: CST.AC_DCC_ACCEPTED_PX,
			value: ['test']
		});
		expect(state).toMatchSnapshot();
	});

	test('dualClassConversions', () => {
		state = dualClassReducer(state, {
			type: CST.AC_DCC_CONVERSIONS,
			value: ['test']
		});
		expect(state).toMatchSnapshot();
	});

	test('dualClassBalances', () => {
		state = dualClassReducer(state, {
			type: CST.AC_DCC_BALANCES,
			value: ['test']
		});
		expect(state).toMatchSnapshot();
	});

	test('dualClassSubscription on', () => {
		state = dualClassReducer(state, {
			type: CST.AC_DCC_SUB,
			custodianType: 'on type',
			tenor: 'on tenor',
			id: 123
		});
		expect(state).toMatchSnapshot();
	});

	test('dualClassSubscription off', () => {
		window.clearInterval = jest.fn();
		state = dualClassReducer(state, {
			type: CST.AC_DCC_SUB,
			custodianType: 'off type',
			tenor: 'off tenor',
			id: 0
		});
		expect(state).toMatchSnapshot();
		expect((window.clearInterval as jest.Mock<Promise<void>>).mock.calls).toMatchSnapshot();
	});
});
