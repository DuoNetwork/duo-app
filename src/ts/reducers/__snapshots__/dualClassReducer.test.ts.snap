// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dualClass reducer default 1`] = `
Object {
  "acceptedPrices": Array [],
  "addresses": Object {
    "aToken": "0x0",
    "bToken": "0x0",
    "feeCollector": "0x0",
    "operator": "0x0",
    "oracle": "0x0",
    "roleManager": "0x0",
  },
  "balances": Object {
    "a": 0,
    "b": 0,
    "duo": 0,
  },
  "conversions": Array [],
  "exchangePrices": Array [],
  "states": Object {
    "alpha": 0,
    "beta": 0,
    "collateral": 0,
    "createCommRate": 0,
    "feeBalance": 0,
    "iterationGasThreshold": 0,
    "lastOperationTime": 0,
    "lastPrice": 0,
    "lastPriceTime": 0,
    "limitLower": 0,
    "limitPeriodic": 0,
    "limitUpper": 0,
    "maturity": 0,
    "minBalance": 0,
    "navA": 0,
    "navB": 0,
    "nextResetAddrIndex": 0,
    "operationCoolDown": 0,
    "period": 0,
    "periodCoupon": 0,
    "preResetWaitingBlocks": 0,
    "priceFetchCoolDown": 0,
    "redeemCommRate": 0,
    "resetPrice": 0,
    "resetPriceTime": 0,
    "resetState": "",
    "state": "Loading",
    "totalSupplyA": 0,
    "totalSupplyB": 0,
    "totalUsers": 0,
  },
  "subscription": 0,
  "tenor": "Perpetual",
  "type": "Beethoven",
}
`;

exports[`dualClass reducer dualClassAcceptedPrices 1`] = `
Object {
  "acceptedPrices": Array [
    "test",
  ],
  "addresses": Object {
    "test": "test",
  },
  "balances": Object {
    "a": 0,
    "b": 0,
    "duo": 0,
  },
  "conversions": Array [],
  "exchangePrices": Object {
    "test": "test",
  },
  "states": Object {
    "test": "test",
  },
  "subscription": 0,
  "tenor": "Perpetual",
  "type": "Beethoven",
}
`;

exports[`dualClass reducer dualClassAddresses 1`] = `
Object {
  "acceptedPrices": Array [],
  "addresses": Object {
    "test": "test",
  },
  "balances": Object {
    "a": 0,
    "b": 0,
    "duo": 0,
  },
  "conversions": Array [],
  "exchangePrices": Array [],
  "states": Object {
    "test": "test",
  },
  "subscription": 0,
  "tenor": "Perpetual",
  "type": "Beethoven",
}
`;

exports[`dualClass reducer dualClassBalances 1`] = `
Object {
  "acceptedPrices": Array [
    "test",
  ],
  "addresses": Object {
    "test": "test",
  },
  "balances": Array [
    "test",
  ],
  "conversions": Array [
    "test",
  ],
  "exchangePrices": Object {
    "test": "test",
  },
  "states": Object {
    "test": "test",
  },
  "subscription": 0,
  "tenor": "Perpetual",
  "type": "Beethoven",
}
`;

exports[`dualClass reducer dualClassConversions 1`] = `
Object {
  "acceptedPrices": Array [
    "test",
  ],
  "addresses": Object {
    "test": "test",
  },
  "balances": Object {
    "a": 0,
    "b": 0,
    "duo": 0,
  },
  "conversions": Array [
    "test",
  ],
  "exchangePrices": Object {
    "test": "test",
  },
  "states": Object {
    "test": "test",
  },
  "subscription": 0,
  "tenor": "Perpetual",
  "type": "Beethoven",
}
`;

exports[`dualClass reducer dualClassExchangePrices 1`] = `
Object {
  "acceptedPrices": Array [],
  "addresses": Object {
    "test": "test",
  },
  "balances": Object {
    "a": 0,
    "b": 0,
    "duo": 0,
  },
  "conversions": Array [],
  "exchangePrices": Object {
    "test": "test",
  },
  "states": Object {
    "test": "test",
  },
  "subscription": 0,
  "tenor": "Perpetual",
  "type": "Beethoven",
}
`;

exports[`dualClass reducer dualClassStates 1`] = `
Object {
  "acceptedPrices": Array [],
  "addresses": Object {
    "aToken": "0x0",
    "bToken": "0x0",
    "feeCollector": "0x0",
    "operator": "0x0",
    "oracle": "0x0",
    "roleManager": "0x0",
  },
  "balances": Object {
    "a": 0,
    "b": 0,
    "duo": 0,
  },
  "conversions": Array [],
  "exchangePrices": Array [],
  "states": Object {
    "test": "test",
  },
  "subscription": 0,
  "tenor": "Perpetual",
  "type": "Beethoven",
}
`;

exports[`dualClass reducer dualClassSubscription off 1`] = `
Object {
  "acceptedPrices": Array [],
  "addresses": Object {
    "aToken": "0x0",
    "bToken": "0x0",
    "feeCollector": "0x0",
    "operator": "0x0",
    "oracle": "0x0",
    "roleManager": "0x0",
  },
  "balances": Object {
    "a": 0,
    "b": 0,
    "duo": 0,
  },
  "conversions": Array [],
  "exchangePrices": Array [],
  "states": Object {
    "alpha": 0,
    "beta": 0,
    "collateral": 0,
    "createCommRate": 0,
    "feeBalance": 0,
    "iterationGasThreshold": 0,
    "lastOperationTime": 0,
    "lastPrice": 0,
    "lastPriceTime": 0,
    "limitLower": 0,
    "limitPeriodic": 0,
    "limitUpper": 0,
    "maturity": 0,
    "minBalance": 0,
    "navA": 0,
    "navB": 0,
    "nextResetAddrIndex": 0,
    "operationCoolDown": 0,
    "period": 0,
    "periodCoupon": 0,
    "preResetWaitingBlocks": 0,
    "priceFetchCoolDown": 0,
    "redeemCommRate": 0,
    "resetPrice": 0,
    "resetPriceTime": 0,
    "resetState": "",
    "state": "Loading",
    "totalSupplyA": 0,
    "totalSupplyB": 0,
    "totalUsers": 0,
  },
  "subscription": 0,
  "tenor": "off tenor",
  "type": "off type",
}
`;

exports[`dualClass reducer dualClassSubscription off 2`] = `
Array [
  Array [
    123,
  ],
]
`;

exports[`dualClass reducer dualClassSubscription on 1`] = `
Object {
  "acceptedPrices": Array [
    "test",
  ],
  "addresses": Object {
    "test": "test",
  },
  "balances": Array [
    "test",
  ],
  "conversions": Array [
    "test",
  ],
  "exchangePrices": Object {
    "test": "test",
  },
  "states": Object {
    "test": "test",
  },
  "subscription": 123,
  "tenor": "on tenor",
  "type": "on type",
}
`;
